gateways/firstdata/firstdata.go

```diff
@@ -0,0 +1,257 @@
+package firstdata
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/BoltApp/sleet"
+	"github.com/BoltApp/sleet/common"
+)
+
+const (
+	authPath = "/payments"
+)
+
+type FirstdataClient struct {
+	host            string
+	apiKey          string
+	apiSecret       string
+	clientRequestID string
+	httpClient      *http.Client
+}
+
+func NewClient(env common.Environment, apiKey, apiSecret string) *FirstdataClient {
+	return NewWithHttpClient(env, apiKey, apiSecret, common.DefaultHttpClient())
+}
+
+func NewWithHttpClient(env common.Environment, apiKey, apiSecret string, httpClient *http.Client) *FirstdataClient {
+	return &FirstdataClient{
+		host:       firstdataHost(env),
+		apiKey:     apiKey,
+		httpClient: httpClient,
+	}
+}
+
+func (client *FirstdataClient) Authorize(request *sleet.AuthorizationRequest) (*sleet.AuthorizationResponse, error) {
+	firstdataAuthRequest, err := buildAuthRequest(request)
+
+	clientRequestId := request.ClientTransactionReference
+
+	if err != nil {
+		return nil, err
+	}
+
+	url := "https://" + client.host + authPath
```

Comment:
even this url building can go into sendRequest

Reply:
Subsequent requests require the transactionReference in the url but the initial auth request does not. So i built the url's outside of send and pass them n as opposed to passing an optional transactionReference variable to the send function

Reply:
Okay sure got it - I realized we did something similar for Cybersource so that's okay. Can we put URLs to the capture/void/refund just so we know how we construct the URLs?